Queque communication

AFH7233 opened the pull request:
In this PR a simple Queue communication is implemented from management to the microservice.

AFH7233 committed:
Home page and create product are added with a proto behaviour

AFH7233 committed:
Configuration for rabbit queues. It lacks of listeners for now.

AFH7233 committed:
RabbitMQ configuration

AFH7233 committed:
Microservice project creation and RabbitMq configuration for managment-api

AFH7233 committed:
Refactor pom arquitecture

AFH7233 committed:
Fix for ProductTransferConfiguration. properties for module crud is added

AFH7233 committed:
Reorganization of the project to add a better pom parent

AFH7233 committed:
Merge branch 'development-completeMVC' into development

AFH7233 committed:
Sending one product trough queue create
Queue configuration was moved to commons project.
Libraries for serialaizing and deserialaizing java 8 date objects added.
Added equal and hashCode method to POJO object.

grunger94 commented on commons-library/pom.xml:
```diff
@@ -8,6 +8,12 @@
+    <parent>
+        <groupId>com.afh.skytouch</groupId>
+        <artifactId>skytouch-test</artifactId>
+        <version>1.0-SNAPSHOT</version>
```
Pasa la version de tu app a una maven property en el parent pom.xml

AFH7233 replied:
Version was removed from the subprojects, but to identify the pom parent it stills needed the pom's version

grunger94 commented on commons-library/src/main/java/com/afh/skytouch/commons/configuration/ProductTransferConfiguration.java:
```diff
@@ -0,0 +1,112 @@
+@EnableRabbit
+public class ProductTransferConfiguration {
+
+    public static final String exchange = "product";
```
Hay que extraer estas constantes a una clase separada que posteriormente pueda ser externalizada con spring cloud configuration

AFH7233 replied:
Fixed

grunger94 commented on commons-library/src/main/java/com/afh/skytouch/commons/configuration/ProductTransferConfiguration.java:
```diff
@@ -0,0 +1,112 @@
+    public static  final String getAll = "product.getAll";
+
+    @Autowired
+    private ConnectionFactory rabbitConnectionFactory;
```
Avoid field injection, favour constructor or setter based injection

AFH7233 replied:
Fixed, it was used setter injection

grunger94 commented on commons-library/src/main/java/com/afh/skytouch/commons/configuration/ProductTransferConfiguration.java:
```diff
@@ -0,0 +1,112 @@
+public class ProductTransferConfiguration {
+
+    public static final String exchange = "product";
+    public static final String create = "product.create";
```
Constants should be uppercased by convention

AFH7233 replied:
fixed

grunger94 commented on commons-library/src/main/java/com/afh/skytouch/commons/configuration/ProductTransferConfiguration.java:
```diff
@@ -0,0 +1,112 @@
+
+    @Bean
+    public Jackson2JsonMessageConverter producerConverter() {
+        Jackson2JsonMessageConverter producerConverter = new Jackson2JsonMessageConverter(mapper());
```
Unnecessary declaration of producerConverter

AFH7233 replied:
Removed

AFH7233 replied:
It was re added, because the application broke without it

grunger94 commented on commons-library/src/main/java/com/afh/skytouch/commons/configuration/ProductTransferConfiguration.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    @Bean
+    public Queue createQueue(){
```
You could use this constructor instead:

public Queue(String name) {
        this(name, true, false, false);
}

grunger94 commented on commons-library/src/main/java/com/afh/skytouch/commons/configuration/ProductTransferConfiguration.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+
+    @Bean(name="createTemplate")
```
Beans are named after the method's name you annotate, any reason to explicitly set it in the annotation?

AFH7233 replied:
I tried that way but didn't found the bean, so I used that method to not lose time

grunger94 commented on commons-library/src/main/java/com/afh/skytouch/commons/configuration/ProductTransferConfiguration.java:
```diff
@@ -0,0 +1,112 @@
+
+    @Bean
+    public ObjectMapper mapper() {
+        ObjectMapper mapper = new ObjectMapper()
```
Variable mapper is redundant

grunger94 commented on commons-library/src/main/java/com/afh/skytouch/commons/dto/GenericProduct.java:
```diff
@@ -10,14 +13,19 @@
     private UUID id;
 
+    @NonNull
     private String Name;
```
Lowercase this variable name

AFH7233 replied:
Fixed

grunger94 commented on management-api/pom.xml:
```diff
@@ -9,28 +9,19 @@
-        <version>1.5.7.RELEASE</version>
+        <groupId>com.afh.skytouch</groupId>
+        <artifactId>skytouch-test</artifactId>
+        <version>1.0-SNAPSHOT</version>
```
Here's where the maven property I talked about before will come in handy

AFH7233 replied:
It didn´t allow me to add a proepertie to the spring boot parent

grunger94 commented on management-api/src/main/java/com/afh/skytouch/managment/controllers/GenericProductController.java:
```diff
@@ -14,20 +16,33 @@
-    @GetMapping("/{uuid}")
-    public ResponseEntity<?> getProduct(@PathVariable UUID uuid){
-        return ResponseEntity.ok().build();
+    private static final String productAtributeName = "product";
```
In the same way as suggested in ProductTransferConfiguration, get these constants out of this class and keep them separate. Uppercase them as well.

grunger94 commented on management-api/src/main/java/com/afh/skytouch/managment/controllers/GenericProductController.java:
```diff
@@ -14,20 +16,33 @@
+    private static final String showProductsPage = "list-products";
+
+    @Autowired
+    ProductSender sender;
```
Implement constructor or setter based injection, also, the field could be private I guess

grunger94 commented on management-api/src/main/java/com/afh/skytouch/managment/queue/producers/ProductSender.java:
```diff
@@ -0,0 +1,23 @@
+public class ProductSender {
+
+    @Autowired
+    @Qualifier("createTemplate")
```
Did you really mean to use the same template for different actions? In that case you're injecting the same bean. Avoid field injection. Check if the field should be private.

grunger94 commented on management-api/src/main/webapp/views/create-product.jsp:
```diff
@@ -1,8 +1,16 @@
 <!DOCTYPE html>
 <%@ taglib prefix="spring" uri="http://www.springframework.org/tags"%>
```
Do you need spring and jstl core libs at all?

AFH7233 replied:
Fixed, removed

grunger94 commented on management-api/src/main/webapp/views/create-product.jsp:
```diff
@@ -1,8 +1,16 @@
+                  action="/genericProduct/add" modelAttribute="product">
+                    <form:label path="id">Id: </form:label><label>"${product.id}"</label><br>
+                    <form:label path="name">Name:</form:label><form:input path="name"/><br>
+                    <form:label path="description">Descripcion:</form:label><form:textarea path="description"/><br>
```
Fix label "descripcion"

AFH7233 replied:
fixed

grunger94 commented on pom.xml:
```diff
@@ -0,0 +1,72 @@
+
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-amqp</artifactId>
```
Are these dependencies needed by all the sub modules?

AFH7233 replied:
They were because of the json converter. They were removed

grunger94 commented on product-crud/src/main/java/com/afh/skytouch/management/queue/listeners/ProductCreationListener.java:
```diff
@@ -0,0 +1,15 @@
+
+    @RabbitListener(queues = ProductTransferConfiguration.create)
+    public void onMessage(GenericProduct product){
+        System.out.println("Resultado: "+product.getId());
```
Avoid using system.out printing, use Logger instead

grunger94 commented on product-crud/src/main/resources/application.properties:
```diff
@@ -0,0 +1 @@
+server.port=8181
```
The exercise description states that .yml should be used instead of .properties

grunger94 replied:
My bad, not a requirement

AFH7233 committed:
Queue configurations moved
Field injection avoided
POM individual version removed
Converter to JSON removed

AFH7233 committed:
Unnecesary imports removed from jsp
Not used libraries for jackson removed

AFH7233 committed:
Removed unused imports

AFH7233 committed:
System.out remmoved it was for quick testing

AFH7233 committed:
Fixed label description

AFH7233 committed:
Field lowerCased

iuresti commented on commons-library/pom.xml:
```diff
@@ -28,10 +33,21 @@
         <dependency>
             <groupId>org.hibernate.javax.persistence</groupId>
             <artifactId>hibernate-jpa-2.1-api</artifactId>
             <version>1.0.2.Final</version>
```
I recommend to move all versions of plugins and dependencies to properties section

AFH7233 replied:
Thanks, those versions were changed to a properties as you suggested

iuresti commented on commons-library/src/main/java/com/afh/skytouch/commons/dto/GenericProduct.java:
```diff
@@ -26,4 +34,18 @@
     }
+
+    @Override
+    public boolean equals(Object object){
```
When you require to overwrite this methods I recommend to let the IDE to create them for you. For example in this case your method does not validate null values... give it a try

AFH7233 replied:
Thanks I will try it!

AFH7233 committed:
Pom reconfiguration

AFH7233 committed:
Converter re added
